# Основы языка Python для аналитиков (семинары). Семинар 1

## Задача 1

На складе лежат разные фрукты в разном количестве.

Нужно написать функцию, которая на вход принимает любое количество названий фруктов и их количество, а возвращает общее количество фруктов на складе

In [1]:
def total_fruits_count(**kwargs)->int:
    result = 0
    for arg in kwargs.values():
         result += arg
    return result

In [2]:
total_fruits_count(apple=5, orange=4, pineapple=3, mango = 2)

14

## Задача 2

Дан список с затратами на рекламу. Но в данных есть ошибки, некоторые затраты имеют отрицательную величину. Удалите такие значения из списка и посчитайте суммарные затраты
[100, 125, -90, 345, 655, -1, 0, 200]
Используйте list comprehensions

In [3]:
advertising_cost_data = [100, 125, -90, 345, 655, -1, 0, 200]

In [4]:
right_cost_data = [i for i in advertising_cost_data if i >= 0]

sum(right_cost_data), right_cost_data

(1425, [100, 125, 345, 655, 0, 200])

Сумма затрат на рекламу после удаления не верных данных = 1425

## Задача 3

Даны два списка.

Дата покупки
['2021-09-14', '2021-12-15', '2021-09-08', '2021-12-05', '2021-10-09', '2021-09-30', '2021-12-22', '2021-11-29', '2021-12-24', '2021-11-26', '2021-10-27', '2021-12-18', '2021-11-09', '2021-11-23', '2021-09-27', '2021-10-02', '2021-12-27', '2021-09-20', '2021-12-13', '2021-11-01', '2021-11-09', '2021-12-06', '2021-12-08', '2021-10-09', '2021-10-31', '2021-09-30', '2021-11-09', '2021-12-13', '2021-10-26', '2021-12-09']

Суммы покупок по датам
[1270, 8413, 9028, 3703, 5739, 4095, 295, 4944, 5723, 3701, 4471, 651, 7037, 4274, 6275, 4988, 6930, 2971, 6592, 2004, 2822, 519, 3406, 2732, 5015, 2008, 316, 6333, 5700, 2887]

**3.1 Найдите, какая выручка у компании в ноябре.**
Используйте list comprehensions

**3.2 Найдите выручку компании в зависимости от месяца.**
Для этого напишите функцию, которая на вход принимает список с датами и список с выручкой, а на выходе словарь, где ключи - это месяцы, а значения - это выручка.
Используйте аннотирование типов.

In [5]:
# Дата покупки
date = ['2021-09-14', '2021-12-15', '2021-09-08', '2021-12-05', '2021-10-09', '2021-09-30', '2021-12-22', '2021-11-29', '2021-12-24', '2021-11-26', '2021-10-27', '2021-12-18', '2021-11-09', '2021-11-23', '2021-09-27', '2021-10-02', '2021-12-27', '2021-09-20', '2021-12-13', '2021-11-01', '2021-11-09', '2021-12-06', '2021-12-08', '2021-10-09', '2021-10-31', '2021-09-30', '2021-11-09', '2021-12-13', '2021-10-26', '2021-12-09']


In [6]:
# Суммы покупок по датам
sums = [1270, 8413, 9028, 3703, 5739, 4095, 295, 4944, 5723, 3701, 4471, 651, 7037, 4274, 6275, 4988, 6930, 2971, 6592, 2004, 2822, 519, 3406, 2732, 5015, 2008, 316, 6333, 5700, 2887]


### 3.1 Решение


In [7]:
len(date), len(sums)

(30, 30)

Размеры списков одинаковывые и равны 30

In [8]:
november_date = { sums[i]:date[i]  for i in range(len(date)) if date[i].split('-')[1] == '10' }
profit_in_november = sum([ sums[i] for i in range(len(date)) if date[i].split('-')[1] == '10'  ])
profit_in_november, november_date

(28645,
 {5739: '2021-10-09',
  4471: '2021-10-27',
  4988: '2021-10-02',
  2732: '2021-10-09',
  5015: '2021-10-31',
  5700: '2021-10-26'})

Выручка у компании в ноябре составила 28645

## 3.2  Решение

In [9]:
def profit_by_month(date_list:list[str], sums_by_date:list[int])->dict:
    temp = {}

    for i in range (len(sums_by_date)):
        if temp.get(date_list[i]):
            val = sums_by_date[i] + temp[date_list[i]]
            temp.update({date_list[i]:val})
        else:
            temp.update({date_list[i]:sums_by_date[i]})

    return temp

In [10]:
result = profit_by_month(date, sums)
result

{'2021-09-14': 1270,
 '2021-12-15': 8413,
 '2021-09-08': 9028,
 '2021-12-05': 3703,
 '2021-10-09': 8471,
 '2021-09-30': 6103,
 '2021-12-22': 295,
 '2021-11-29': 4944,
 '2021-12-24': 5723,
 '2021-11-26': 3701,
 '2021-10-27': 4471,
 '2021-12-18': 651,
 '2021-11-09': 10175,
 '2021-11-23': 4274,
 '2021-09-27': 6275,
 '2021-10-02': 4988,
 '2021-12-27': 6930,
 '2021-09-20': 2971,
 '2021-12-13': 12925,
 '2021-11-01': 2004,
 '2021-12-06': 519,
 '2021-12-08': 3406,
 '2021-10-31': 5015,
 '2021-10-26': 5700,
 '2021-12-09': 2887}